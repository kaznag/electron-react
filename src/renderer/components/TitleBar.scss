
$title-bar-height: 30px;
$drag-region-margin: 4px;
$title-bar-button-width: 45px;

.title-bar {
  width: 100%;
  height: $title-bar-height;
}

.title-bar.window-title {
  font-size: 12px;
  text-align: center;
  line-height: $title-bar-height;
}

.title-bar.button {
  float: right;
  width: $title-bar-button-width;
  height: 100%;
}

.title-bar.drag-region {
  position: absolute;
  left: $drag-region-margin;
  top: $drag-region-margin;
  height: calc(#{$title-bar-height} - #{$drag-region-margin});
  width: calc(100% - #{$drag-region-margin} - #{$title-bar-button-width} * 3);
  -webkit-app-region: drag;
}

.title-bar.button.close {
  position: relative;
  background-color: transparent;
}

.title-bar.button.close:before, .title-bar.title-bar.button.close:after {
  position: absolute;
  left: 16px;
  width: 13px;
  height: 1px;
  top: 14px;
  content: " ";
  background-color: #000000;
}

.title-bar.button.close:before {
  transform: rotate(45deg);
}

.title-bar.button.close:after {
  transform: rotate(-45deg);
}

.title-bar.button.close:hover {
  background-color: #e81123;
}

.title-bar.button.close:hover:before, .title-bar.button.close:hover:after {
  position: absolute;
  left: 16px;
  width: 13px;
  height: 1px;
  top: 14px;
  content: " ";
  background-color: #ffffff;
}

.title-bar.button.close:hover:before {
  transform: rotate(45deg);
}

.title-bar.button.close:hover:after {
  transform: rotate(-45deg);
}

.title-bar.button.maximize {
  position: relative;
  background-color: transparent;
}

.title-bar.button.maximize:before {
  position: absolute;
  top: 10px;
  left: 17px;
  width: 8px;
  height: 8px;
  content: " ";
  border: 1px solid #000000;
}

.title-bar.button.maximize:hover {
  background-color: #e5e5e5;
}

.title-bar.button.restore {
  position: relative;
  background-color: transparent;
}

.title-bar.button.restore:before {
  position: absolute;
  top: 12px;
  left: 17px;
  width: 6px;
  height: 6px;
  content: " ";
  border: 1px solid #000000;
}

.title-bar.button.restore:after {
  position: absolute;
  top: 10px;
  left: 19px;
  width: 1px;
  height: 1px;
  content: " ";
  background: black;
  box-shadow:
    0px 1px,
    1px 0px,
    2px 0px,
    3px 0px,
    4px 0px,
    5px 0px,
    6px 0px,
    7px 0px,
    7px 1px,
    7px 2px,
    7px 3px,
    7px 4px,
    7px 5px,
    7px 6px,
    7px 7px,
    6px 7px,
}

.title-bar.button.restore:hover {
  background-color: #e5e5e5;
}

.title-bar.button.minimize {
  position: relative;
  background-color: transparent;
}

.title-bar.button.minimize:before {
  position: absolute;
  top: 15px;
  left: 17px;
  width: 10px;
  height: 1px;
  content: " ";
  border-top: 1px solid #000000;
}

.title-bar.button.minimize:hover {
  background-color: #e5e5e5;
}
