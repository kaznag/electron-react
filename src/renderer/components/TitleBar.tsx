import React from 'react';

import './TitleBar.scss';

type TitleBarProps = {
  isFocused: boolean;
  isMaximized: boolean;
  windowTitle?: string;
  onCloseButtonClick?: (event: React.MouseEvent<HTMLElement>) => void;
  onMaximizeRestoreButtonClick?: (event: React.MouseEvent<HTMLElement>) => void;
  onMinimizeButtonClick?: (event: React.MouseEvent<HTMLElement>) => void;
}

const TitleBar: React.FC<TitleBarProps> = props => {
  return (
    <div className="title-bar">
      <div className="drag-region"></div>
      <div className={'close-button' + (!props.isFocused ? ' blur' : '')} onClick={props.onCloseButtonClick}></div>
      <div className={(props.isMaximized ? 'title-bar button restore' : 'maximize-button') + (!props.isFocused ? ' blur' : '')} onClick={props.onMaximizeRestoreButtonClick}></div>
      <div className="title-bar button minimize" onClick={props.onMinimizeButtonClick}></div>
      <div className={'window-title' + (!props.isFocused ? ' blur' : '')}>{props.windowTitle}</div>
    </div>
  );
};

export default TitleBar;
