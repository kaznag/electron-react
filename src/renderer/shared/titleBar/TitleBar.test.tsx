import '@testing-library/jest-dom';

import { configure, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { TitleBar } from './TitleBar';

configure({ testIdAttribute: 'data-test-id' });

describe('<TitleBar />', () => {
  test('should render default TitleBar', () => {
    const { asFragment } = render(<TitleBar />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render with resize button', () => {
    const { asFragment } = render(<TitleBar isMaximized={true} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render blur TitleBar', () => {
    const { asFragment } = render(<TitleBar isFocused={false} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render with blur resize button', () => {
    const { asFragment } = render(<TitleBar isFocused={false} isMaximized={true} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should render title', () => {
    const { asFragment } = render(<TitleBar windowTitle={'Hello world'} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should click close button', () => {
    const handler = jest.fn();
    const { queryByTestId, getByTestId } = render(<TitleBar onCloseButtonClick={handler} />);

    expect(queryByTestId('close-button')).toBeTruthy();

    fireEvent.click(getByTestId('close-button'));
    expect(handler).toHaveBeenCalled();
  });

  test('should click maximize button', () => {
    const handler = jest.fn();
    const { queryByTestId, getByTestId } = render(
      <TitleBar onMaximizeResizeButtonClick={handler} />
    );

    expect(queryByTestId('maximize-button')).toBeTruthy();

    fireEvent.click(getByTestId('maximize-button'));
    expect(handler).toHaveBeenCalled();
  });

  test('should click minimize button', () => {
    const handler = jest.fn();
    const { queryByTestId, getByTestId } = render(<TitleBar onMinimizeButtonClick={handler} />);

    expect(queryByTestId('minimize-button')).toBeTruthy();

    fireEvent.click(getByTestId('minimize-button'));
    expect(handler).toHaveBeenCalled();
  });
});
