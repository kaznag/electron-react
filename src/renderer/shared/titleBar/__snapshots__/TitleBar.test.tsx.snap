// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TitleBar /> should render blur TitleBar 1`] = `
<DocumentFragment>
  <div
    class="title-bar"
  >
    <div
      class="drag-region"
    />
    <div
      class="close-button blur"
      data-test-id="close-button"
    />
    <div
      class="maximize-button blur"
      data-test-id="maximize-button"
    />
    <div
      class="minimize-button blur"
      data-test-id="minimize-button"
    />
    <div
      class="window-title blur"
      data-test-id="window-title"
    />
  </div>
</DocumentFragment>
`;

exports[`<TitleBar /> should render default TitleBar 1`] = `
<DocumentFragment>
  <div
    class="title-bar"
  >
    <div
      class="drag-region"
    />
    <div
      class="close-button"
      data-test-id="close-button"
    />
    <div
      class="maximize-button"
      data-test-id="maximize-button"
    />
    <div
      class="minimize-button"
      data-test-id="minimize-button"
    />
    <div
      class="window-title"
      data-test-id="window-title"
    />
  </div>
</DocumentFragment>
`;

exports[`<TitleBar /> should render title 1`] = `
<DocumentFragment>
  <div
    class="title-bar"
  >
    <div
      class="drag-region"
    />
    <div
      class="close-button"
      data-test-id="close-button"
    />
    <div
      class="maximize-button"
      data-test-id="maximize-button"
    />
    <div
      class="minimize-button"
      data-test-id="minimize-button"
    />
    <div
      class="window-title"
      data-test-id="window-title"
    >
      Hello world
    </div>
  </div>
</DocumentFragment>
`;

exports[`<TitleBar /> should render with blur resize button 1`] = `
<DocumentFragment>
  <div
    class="title-bar"
  >
    <div
      class="drag-region"
    />
    <div
      class="close-button blur"
      data-test-id="close-button"
    />
    <div
      class="resize-button blur"
      data-test-id="maximize-button"
    />
    <div
      class="minimize-button blur"
      data-test-id="minimize-button"
    />
    <div
      class="window-title blur"
      data-test-id="window-title"
    />
  </div>
</DocumentFragment>
`;

exports[`<TitleBar /> should render with resize button 1`] = `
<DocumentFragment>
  <div
    class="title-bar"
  >
    <div
      class="drag-region"
    />
    <div
      class="close-button"
      data-test-id="close-button"
    />
    <div
      class="resize-button"
      data-test-id="maximize-button"
    />
    <div
      class="minimize-button"
      data-test-id="minimize-button"
    />
    <div
      class="window-title"
      data-test-id="window-title"
    />
  </div>
</DocumentFragment>
`;
